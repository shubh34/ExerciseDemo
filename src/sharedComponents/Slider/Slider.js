/* eslint-disable no-return-assign */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './Slider.scss';
import SliderWrapper from './SliderWrapper/SliderWrapper';
import SlideButton from './SlideButton/SlideButton';

class Slider extends Component {
	constructor(props) {
		super(props);
		this.handleNext = this.handleNext.bind(this);
		this.handlePrevious = this.handlePrevious.bind(this);
		this.state = {
			distance: 0,
			hasNext: true,
			hasPrev: false
		};
	}

	handlePrevious() {
		const { distance } = this.state;
		this.setState(
			{
				distance: distance + this.silder.clientWidth
			},
			() => {
				const { distance } = this.state;
				this.setState({
					hasNext: this.silder.scrollWidth - Math.abs(distance) > 0,
					hasPrev: distance < 0
				});
			}
		);
	}

	handleNext() {
		const { distance } = this.state;
		this.setState(
			{
				distance: distance - this.silder.clientWidth
			},
			() => {
				const { distance } = this.state;
				this.setState({
					hasNext: this.silder.scrollWidth - Math.abs(distance) > this.silder.clientWidth,
					hasPrev: distance < 0
				});
			}
		);
	}

	render() {
		const { distance, hasPrev, hasNext } = this.state;
		const { children } = this.props;
		const slideStyle = {
			transform: `translate3d(${distance}px, 0, 0)`
		};
		return (
			<SliderWrapper>
				<div className='slider'>
					<div ref={ref => (this.silder = ref)} style={slideStyle} className='slider__container'>
						{children}
					</div>
				</div>
				{hasPrev && <SlideButton type='prev' onClick={this.handlePrevious} />}
				{hasNext && <SlideButton type='next' onClick={this.handleNext} />}
			</SliderWrapper>
		);
	}
}

Slider.propTypes = {
	children: PropTypes.arrayOf(PropTypes.element).isRequired
};

export default React.memo(Slider);
