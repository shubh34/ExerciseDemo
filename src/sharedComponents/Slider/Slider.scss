
.slider {
    display: flex;
    position: relative;
  
    &__container {
      display: flex;
      transition: transform 300ms ease 100ms;
      z-index: 3;
      width: 100%;
    }
  }
  