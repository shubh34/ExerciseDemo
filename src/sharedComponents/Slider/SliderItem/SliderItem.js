import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import cx from 'classnames';
import './SliderItem.scss';
import Scores from '../../Scores/Scores';

const SliderItem = ({
	movie, id, score, votes,
}) => (
	<div
		className={cx('slider-item')}
	>
		<Link className="slider-item-image" to={`/movies/${id}`}>
			<img width="100%" src={movie} alt="" />
		</Link>
		<Scores score={score} votes={votes} />
	</div>
);

SliderItem.propTypes = {
	movie: PropTypes.string.isRequired,
	id: PropTypes.string.isRequired,
	score: PropTypes.number.isRequired,
	votes: PropTypes.string.isRequired,
};

export default SliderItem;
