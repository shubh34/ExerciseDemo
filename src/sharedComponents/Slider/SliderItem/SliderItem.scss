.slider-item {
    flex: 0 0 35.5%;
    padding: 0 5px;
    position: relative;
    border-radius: 2px;
    @media screen and (min-width: 560px) {
      flex: 0 0 22.5%;
    }
    @media screen and (min-width: 1140px) {
      flex: 0 0 15.5%;
    }
    @media screen and (min-width: 1400px) {
      flex: 0 0 13.5%;
    }

  
   
  }
  .slider-item-image {
    vertical-align: top;
  }
  .slider-item-image :hover {
    border: 3px #ffe300 solid;
    padding: 0px;
    cursor: pointer;
    border-radius: 5px; 
  }
