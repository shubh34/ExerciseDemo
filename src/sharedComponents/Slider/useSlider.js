import { useState, useEffect } from 'react';

export const getMiddleList = (defaultArray, middleIndex, defaultElementSizing) => {
	const actualArrayLength = defaultArray.length;
	const lastIndex = middleIndex + defaultElementSizing;
	const actualLastIndex = lastIndex % actualArrayLength;
	if (lastIndex >= actualArrayLength) {
		const middleArray = defaultArray.slice(middleIndex);
		return [...middleArray, ...defaultArray.slice(0, actualLastIndex)];
	}

	return defaultArray.slice(middleIndex, lastIndex);
};

export const getLastList = (defaultArray, middleIndex, defaultElementSizing) => {
	const actualArrayLength = defaultArray.length;
	const lastIndex = middleIndex + defaultElementSizing;
	const actualLastIndex = lastIndex % actualArrayLength;
	const lastArray = defaultArray.slice(actualLastIndex, actualLastIndex + defaultElementSizing);
	if (actualLastIndex + defaultElementSizing >= actualArrayLength) {
		return [...lastArray, ...defaultArray.slice(0, (actualLastIndex + defaultElementSizing + 1) % actualArrayLength)];
	}
	return lastArray;
};
export const getFirstList = (defaultArray, middleIndex, defaultElementSizing) => {
	const actualArrayLength = defaultArray.length;
	const firstIndex = middleIndex - defaultElementSizing;
	const actualFirstIndex = firstIndex >= 0 ? firstIndex : actualArrayLength - Math.abs(firstIndex);
	const firstElementIndex = actualFirstIndex === 0 ? actualArrayLength - 1 : actualFirstIndex - 1;
	const firstElement = defaultArray[firstElementIndex];
	if (actualFirstIndex + defaultElementSizing >= defaultArray.length) {
		const firstArray = defaultArray.slice(actualFirstIndex);
		return [firstElement, ...firstArray, ...defaultArray.slice(0, (actualFirstIndex + defaultElementSizing) % actualArrayLength)];
	}
	return [firstElement, ...defaultArray.slice(actualFirstIndex, (actualFirstIndex + defaultElementSizing) % actualArrayLength)];
};

export const getSliderElement = (defaultArray, middleIndex, defaultElementSizing) => {
	const firstList = getFirstList(defaultArray, middleIndex, defaultElementSizing);
	const middleList = getMiddleList(defaultArray, middleIndex, defaultElementSizing);
	const lastList = getLastList(defaultArray, middleIndex, defaultElementSizing);
	return [...firstList, middleList, lastList];
};

export const useSlider = (sliderRef, actualSliderItems, defaultElementSizing = 7) => {
	const [sliderItems, setSliderItems] = useState(actualSliderItems);
	const [transition, setSliderTransition] = useState(true);
	const [activeSliderIndex, setActiveSliderItemIndex] = useState(0);
	const [translate, setSliderTranslate] = useState(0);
	const handlePrevious = () => {
		const calculatedActiveIndex = activeSliderIndex - defaultElementSizing;
		const actualActiveIndex = (calculatedActiveIndex >= 0 ? calculatedActiveIndex : defaultElementSizing - Math.abs(calculatedActiveIndex)) % actualSliderItems.length;
		setSliderTranslate(translate + 1.13);
		setSliderTransition(true);
		setTimeout(() => {
			setSliderItems(getSliderElement(actualSliderItems, actualActiveIndex, defaultElementSizing));
			setActiveSliderItemIndex(actualActiveIndex);
			setSliderTranslate(-1.13);
		}, 100);
	};
	const handleNext = () => {
		const calculatedActiveIndex = activeSliderIndex + defaultElementSizing;
		const actualActiveIndex = calculatedActiveIndex % actualSliderItems.length;
		setSliderTranslate(translate - 1.1425);
		setSliderTransition(true);
		setTimeout(() => {
			setSliderItems(getSliderElement(actualSliderItems, actualActiveIndex, defaultElementSizing));
			setActiveSliderItemIndex(actualActiveIndex);
			setSliderTransition(false);
			setSliderTranslate(-1.145);
		}, 300);
	};
	useEffect(() => {
		setSliderItems(actualSliderItems);
	}, [actualSliderItems]);
	return { transition, translate, sliderItems, handlePrevious, handleNext };
};
