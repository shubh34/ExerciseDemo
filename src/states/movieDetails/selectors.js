import get from 'lodash/get';
import { createSelector } from 'reselect';

const getMovieDetails = state => get(state, 'movieDetails', '');

export const getHighLightedScore = createSelector(getMovieDetails, movieDetails => get(movieDetails, 'data.scores', []).find(score => score.highlighted));

export const getMovieTitle = createSelector(getMovieDetails, movieDetails => get(movieDetails, 'data.title', ''));

export const getSnapshotImageUrl = createSelector(getMovieDetails, movieDetails => get(movieDetails, 'data.images.snapshot', '').replace('.jpeg', '-width1920-quality80.jpeg'));

export const getMovieDescription = createSelector(getMovieDetails, movieDetails => get(movieDetails, 'data.plot', ''));

export const getScore = createSelector(getHighLightedScore, scores => get(scores, 'score', ''));

export const getVotes = createSelector(getHighLightedScore, scores => get(scores, 'formatted_amount_of_votes', ''));
