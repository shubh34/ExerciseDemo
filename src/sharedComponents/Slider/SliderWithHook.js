/* eslint-disable no-return-assign */
import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import './Slider.scss';
import SliderWrapper from './SliderWrapper/SliderWrapper';
import SlideButton from './SlideButton/SlideButton';
import { useSlider } from './useSlider';

const SliderWithHook = ({ children }) => {
	const sliderRef = useRef();
	const { transition, translate, handlePrevious, handleNext, sliderItems } = useSlider(sliderRef, children);
	const slideStyle = {
		transform: `translate3d(${translate * 100}%, 0, 0)`,
		transition: transition ? 'transform .5s ease 0s' : 'none'
	};
	return (
		<SliderWrapper>
			<div className='slider'>
				<div ref={sliderRef} style={slideStyle} className='slider__container'>
					{sliderItems}
				</div>
			</div>
			<SlideButton type='prev' onClick={handlePrevious} />
			<SlideButton type='next' onClick={handleNext} />
		</SliderWrapper>
	);
};

SliderWithHook.propTypes = {
	children: PropTypes.arrayOf(PropTypes.element).isRequired
};

export default SliderWithHook;
