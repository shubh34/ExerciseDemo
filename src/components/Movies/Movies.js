import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { fetchMovieListByCategory } from '../../states/movieListing/action';
import SectionHeader from '../../sharedComponents/SectionHeader/SectionHeader';
import Slider from '../../sharedComponents/Slider/Slider';
import SliderItem from '../../sharedComponents/Slider/SliderItem/SliderItem';
import { getMovieListHeader, getMovieListByCategory } from '../../states/movieListing/selectors';


const mapDispatch = dispatch => ({
	fetchMovieList: id => dispatch(fetchMovieListByCategory(id)),
});
const mapState = (state, props) => ({
	movieCategoryHeader: getMovieListHeader(state, props.id),
	movieList: getMovieListByCategory(state, props.id),
});
export const Movies = (props) => {
	const {
		id, fetchMovieList, movieCategoryHeader, movieList,
	} = props;
	useEffect(() => {
		fetchMovieList(id);
	}, [id]);
	return (
		<div className="movies">
			<SectionHeader header={movieCategoryHeader} />
			<Slider>
				{movieList.map(
					({
						artwork, score, id, votes,
					}, index) => (
						<SliderItem movie={artwork} score={score} id={id} votes={votes} key={index} />
					),
				)}
			</Slider>
		</div>
	);
};

Movies.propTypes = {
	id: PropTypes.string.isRequired,
	movieCategoryHeader: PropTypes.string.isRequired,
	movieList: PropTypes.arrayOf(PropTypes.shape({
		artwork: PropTypes.string.isRequired,
		score: PropTypes.number.isRequired,
		id: PropTypes.string.isRequired,
		votes: PropTypes.string.isRequired,
	})),
	fetchMovieList: PropTypes.func.isRequired,
};
export default connect(mapState, mapDispatch)(Movies);
