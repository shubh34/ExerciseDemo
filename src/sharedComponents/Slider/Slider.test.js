
import React from 'react';
import { shallow } from 'enzyme';

import { rawMovies } from '../../../tests-stub/Movies';
import { getMovieListByCategory } from '../../states/movieListing/selectors';
import SliderItem from './SliderItem/SliderItem';
import Slider from './Slider';


const setUp = (container = shallow) => {
	const state = { movieListing: rawMovies };
	const movieList = getMovieListByCategory(state, 'populares-en-taquilla');

	const wrapper = container(
		<Slider>
			{movieList.map(
				({
					artwork, score, id, votes,
				}, index) => (
					<SliderItem movie={artwork} score={score} id={id} votes={votes} key={index} ref={React.createRef()} />
				),
			)}
		</Slider>,
	);
	return { wrapper };
};

describe('Slider', () => {
	it('should render Movies in Sliders', () => {
		const { wrapper } = setUp();
		expect(wrapper).toMatchSnapshot();
	});
});
