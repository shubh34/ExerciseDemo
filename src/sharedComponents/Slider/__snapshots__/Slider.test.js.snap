// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should render Movies in Sliders 1`] = `
<SliderWrapper>
  <div
    className="slider"
  >
    <div
      className="slider__container"
      style={
        Object {
          "transform": "translate3d(0px, 0, 0)",
        }
      }
    >
      <SliderItem
        id="capitana-marvel"
        key="0"
        movie="https://images-2.wuaki.tv/system/artworks/86665/master/capitana-marvel-1560352921-width217-quality80.jpeg"
        score={7}
        votes="6,5K"
      />
      <SliderItem
        id="a-47-metros"
        key="1"
        movie="https://images-3.wuaki.tv/system/artworks/54649/master/a-47-metros-1558460397-width217-quality80.jpeg"
        score={5.6}
        votes="1,6K"
      />
      <SliderItem
        id="green-book"
        key="2"
        movie="https://images-2.wuaki.tv/system/artworks/86659/master/green-book-1558483188-width217-quality80.jpeg"
        score={8.3}
        votes="4K"
      />
      <SliderItem
        id="mascotas"
        key="3"
        movie="https://images-2.wuaki.tv/system/artworks/43779/master/mascotas-1558452275-width217-quality80.jpeg"
        score={6.1}
        votes="5,6K"
      />
      <SliderItem
        id="mula"
        key="4"
        movie="https://images-2.wuaki.tv/system/artworks/89878/master/mula-1558485497-width217-quality80.jpeg"
        score={6.5}
        votes="1,8K"
      />
      <SliderItem
        id="escape-room-2019"
        key="5"
        movie="https://images-0.wuaki.tv/system/artworks/89879/master/escape-room-2019-1559899512-width217-quality80.jpeg"
        score={6.4}
        votes="1,5K"
      />
      <SliderItem
        id="perdiendo-el-este"
        key="6"
        movie="https://images-1.wuaki.tv/system/artworks/86635/master/perdiendo-el-este-1559228752-width217-quality80.jpeg"
        score={4.2}
        votes="20"
      />
      <SliderItem
        id="creed-ii-la-leyenda-de-rocky"
        key="7"
        movie="https://images-3.wuaki.tv/system/artworks/79289/master/creed-ii-la-leyenda-de-rocky-1558478964-width217-quality80.jpeg"
        score={6.7}
        votes="2,3K"
      />
    </div>
  </div>
  <SlideButton
    onClick={[Function]}
    type="next"
  />
</SliderWrapper>
`;
